hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,180 @@
+            if total_msg > 0:
+                cov_rate = fpv_summary["covered"] / total_msg * 100
+            else:
+                cov_rate = 0
+            fpv_summary["cov_rate"] = "{0:.2f} %".format(round(cov_rate, 2))
+
+    except IOError as err:
+        messages["errors"] += ["IOError: %s" % err]
+
+    results["messages"] = messages
+    if fpv_summary:
+        results["fpv_summary"] = fpv_summary
+
+    return results
+
+
+def get_cov_results(resdir):
+    """
+    Parse FPV coverage information from the log file.
+    If no coverage information found, the function will return None
+    """
+    cov_results = OrderedDict()
+    try:
+        with Path(resdir).joinpath('fpv.log').open() as f:
+            full_file = f.read()
+            coverage_patterns = [("stimuli", r"^\|\w+.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*\|\d*.\d\d%.*"),
+                                 ("coi", r"^\|\w+.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*"),
+                                 ("proof", r"^\|\w+.*\|\d*.\d\d%.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*")]
+            extract_messages(full_file, coverage_patterns, cov_results)
+            for key, item in cov_results.items():
+                if len(item) == 1:
+                    cov_results[key] = item[0] + " %"
+                else:
+                    cov_results = None
+                    break
+
+    except IOError as err:
+        print(err)
+        cov_results = None
+
+    return cov_results
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=
+        """This script parses The JasperGold FPV log to extract below information,
+        "messages": {
+          "errors"      : []
+          "warnings"    : []
+          "fpv_errors"  : []
+          "fpv_warnings": []
+          "cex"         : ["property1", "property2"...],
+          "undetermined": [],
+          "unreachable" : [],
+        },
+        "fpv_summary": {
+          "fpv_errors"  : 0
+          "fpv_warnings": 2
+          "proven"      : 20,
+          "cex"         : 5,
+          "covered"     : 18,
+          "undetermined": 7,
+          "unreachable" : 2,
+          "pass_rate"   : "90 %",
+          "cover_rate"  : "90 %"
+        },
+        If coverage is enabled, this script will also generate a coverage summary:
+        "fpv_coverage": {
+          stimuli: "90 %",
+          coi    : "90 %",
+          proof  : "80 %"
+        }
+        """)
+    parser.add_argument('--repdir',
+                        type=str,
+                        default="",
+                        help="""The script searches the 'fpv.log' under the repdir.
+                             Defaults to the current dirctory""")
+
+    parser.add_argument('--outdir',
+                        type=str,
+                        default="",
+                        help="""Output directory for the parsed 'results.hjson' file.
+                             Defaults to the current dirctory""")
```

Comment:
Wouldn't it be cleaner to pass the input and output file paths, rather than directories in which to find files with a special name?